doc/patch-author-guide.md

```diff
@@ -597,3 +579,126 @@ For example, a `pr_warn_once()` can be replaced with:
+```
+
+3. a peek at the generated code:
+
+```
+Disassembly of section .text.__tcp_mtu_to_mss:
+
+0000000000000000 <__tcp_mtu_to_mss>:
+   0:   48 8b 87 60 05 00 00    mov    0x560(%rdi),%rax
+   7:   0f b7 50 30             movzwl 0x30(%rax),%edx
+   b:   0f b7 40 32             movzwl 0x32(%rax),%eax
+   f:   29 d6                   sub    %edx,%esi
+  11:   83 ee 14                sub    $0x14,%esi
+  ...
+```
+
+This could be a little confusing since one might have expected to see
+changes to all of `__tcp_mtu_to_mss()` callers (ie, it was inlined as
+requested).  In this case, a simple workaround is to provide a new,
+non-inlined copy of `__tcp_mtu_to_mss()` and then update all its callers
+to use it.
+
+Jump labels
+-----------
+
+When modifying a function that contains a jump label, kpatch-build may
+return an error like: `ERROR: oom_kill.o: kpatch_regenerate_special_section: 2109: Found a jump label at out_of_memory()+0x10a, using key cpusets_enabled_key.  Jump labels aren't currently supported.  Use static_key_enabled() instead.`
+
+This is due to a limitation in the kernel to process static key
+livepatch relocations (resolved by late-module patching).  Older
+versions of kpatch-build may have reported successfully building 
+kpatch module, but issue
+[#931](https://github.com/dynup/kpatch/issues/931) revealed potentially
+dangerous behavior if the static key value had been modified from its
+compiled default.
+
+The current workaround is to remove the jump label by explictly checking
+the static key:
+
+```
+/* unsupported */
+if (static_key_true(&key1))
+
+/* supported */
+if (static_key_enabled(&key1))
+```
+
+Constant propagation symbol changes
+-----------------------------------
+
+GCC's constant propagation optimizations may result in kpatch-build
+errors like: `ERROR: timekeeping.o: symbol changed sections: .text.timekeeping_update.constprop.9`
+
+This is currently tracked as issue
+[#935](https://github.com/dynup/kpatch/issues/935) "duplicate" mangled
+fuctions.  In that case, readelf revealed that kpatch-build wasn't
+correleating constprop subfunctions created by the optimization:
+```
+# eu-readelf -s orig/kernel/time/timekeeping.o  | grep timekeeping_update
+  137: 0000000000000000    552 FUNC    LOCAL  DEFAULT       48 timekeeping_update.constprop.9
+  138: 0000000000000000    568 FUNC    LOCAL  DEFAULT       50 timekeeping_update.constprop.8
+# eu-readelf -s patched/kernel/time/timekeeping.o  | grep timekeeping_update
+  137: 0000000000000000    552 FUNC    LOCAL  DEFAULT       48 timekeeping_update.constprop.10
+  138: 0000000000000000    568 FUNC    LOCAL  DEFAULT       50 timekeeping_update.constprop.9
+```
+The current workaround for this issue to add `__attribute__((noclone))`
+to offending functions.
+
+Sibling calls
+-------------
+
+If function A() calls function B() and they share a common function
+signature, GCC may deem them "sibling calls" and apply a tail call
```

Comment:
Hmm, I suppose that is true, I haven't looked into the implementation of how they implement it.  As long as the stack expectations line up when B() is called, I can see how the signatures don't need to match exactly.  I can reword this to state that the compiler may optimize A() calling B() at the end of the function, skipping some calling conventions to save overhead.

Reply:
The signatures don't need to match because for sibling calls, A() restores the stack to its callee's state, and then sets up B()'s args appropriately before jumping to B().